clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,390 @@
+     * get_filetype() API on a \p memtrace_stream_t. However, not all instances of
+     * \p memtrace_stream_t have the filetype available at init time (before the
+     * \p TRACE_MARKER_TYPE_FILETYPE is actually returned by the stream).
+     * Particularly, the \p stream_t provided by the \p scheduler_t to analysis
+     * tools provides the file type at init time when #dynamorio::drmemtrace::
+     * scheduler_tmpl_t::scheduler_options_t.read_inputs_in_init is set. In other
+     * cases, the user may call this API after init time when the
+     * \p TRACE_MARKER_TYPE_FILETYPE marker is seen (still before any instr
+     * record).
+     *
+     * If the provided \p module_file_path is empty, the cache object uses
+     * the encodings embedded in the trace records.
+     *
+     * If the \p module_file_path parameter is not empty, it instructs the
+     * \p decode_cache_t object that it should look for the instr encodings in the
+     * application binaries using a \p module_mapper_t. A \p module_mapper_t is
+     * instantiated only one time and reused for all objects
+     * of \p decode_cache_t (of any template type). The user must provide a valid
+     * \p module_file_path if decoding instructions from a trace that does not
+     * have embedded instruction encodings in it, as indicated by absence of
+     * the \p OFFLINE_FILE_TYPE_ENCODINGS bit in the \p TRACE_MARKER_TYPE_FILETYPE
+     * marker. The user may provide a \p module_file_path also if they
+     * deliberately need to use the module mapper instead of the embedded
+     * encodings.
+     */
+    std::string
+    init(offline_file_type_t filetype, const std::string &module_file_path = "",
+         const std::string &alt_module_dir = "")
+    {
+        // TODO i#7113: Should we perhaps also do dr_set_isa_mode for DR_ISA_REGDEPS
+        // that cannot be figured out automatically using the build but needs the
+        // trace filetype? Or does that belong in the analyzer framework which is more
+        // central than this decode_cache_t.
+        init_done_ = true;
+        if (!TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, filetype) && module_file_path.empty()) {
+            return "Trace does not have embedded encodings, and no module_file_path "
+                   "provided";
+        }
+        if (module_file_path.empty())
+            return "";
+        return init_module_mapper(module_file_path, alt_module_dir);
+    }
+
+private:
+    std::unordered_map<app_pc, DecodeInfo> decode_cache_;
+    void *dcontext_ = nullptr;
+    bool persist_decoded_instrs_ = false;
+
+    // Describes whether init() was invoked.
+    // This helps in detecting cases where a module mapper was not specified
+    // when decoding a trace without embedded encodings.
+    bool init_done_ = false;
+};
+
+/**
+ *  An \p decode_cache_t for testing which uses a \p test_module_mapper_t.
+ */
+template <class DecodeInfo>
+class test_decode_cache_t : public decode_cache_t<DecodeInfo> {
```

Comment:
Does this need to be in this header?  Shouldn't this be in the test file?

Reply:
Seemed convenient to have the test class available at a common location. But sure, it'll be better to find a new test-only common location.

Reply:
Actually test_module_mapper_t also lives in raw2trace_shared.h. Do we want to update all such test classes to be in their own separate header?